tests/unit/lms/services/jwt_test.py

```diff
@@ -57,24 +63,112 @@ def test_encode_with_secret(self, svc, jwt):
         )
         assert encoded_jwt == jwt.encode.return_value
 
+    def test_decode_lti_token(self, svc, jwt, lti_registration_service):
+        registration = factories.LTIRegistration(key_set_url="http://jwk.com")
+        lti_registration_service.get.return_value = registration
+        jwt.decode.return_value = {"aud": "AUD", "iss": "ISS"}
+
+        payload = svc.decode_lti_token(sentinel.id_token)
+
+        lti_registration_service.get.assert_called_once_with("ISS", "AUD")
+        jwt.PyJWKClient.assert_called_once_with(
+            lti_registration_service.get.return_value.key_set_url
+        )
+        jwt.decode.assert_called_with(
+            sentinel.id_token,
+            key=jwt.PyJWKClient.return_value.get_signing_key_from_jwt.return_value.key,
+            audience="AUD",
+            algorithms=["RS256"],
+        )
+        assert payload == {"aud": "AUD", "iss": "ISS"}
+
+    def test_decode_lti_token_with_empty_token(self, svc):
+        assert not svc.decode_lti_token("")
+
+    def test_decode_lti_token_with_no_kid(self, svc):
+        jwt_str = self.encode_jwt({"key": "value"}, headers={"key": "value"})
+
+        assert not svc.decode_lti_token(jwt_str)
+
+    def test_decode_lti_token_with_no_registration(
+        self, svc, jwt, lti_registration_service
+    ):
+        jwt.decode.return_value = {"aud": "AUD", "iss": "ISS"}
+        lti_registration_service.get.return_value = None
+
+        assert not svc.decode_lti_token(sentinel.id_token)
+
+    def test_decode_lti_token_with_invalid_jwt(self, svc, jwt):
+        jwt.decode.side_effect = [{"aud": "AUD", "iss": "ISS"}, InvalidTokenError()]
+
+        assert not svc.decode_lti_token(sentinel.id_token)
+
+    @staticmethod
     def encode_jwt(
-        self,
         payload,
-        lifetime,
         secret="test_secret",
         algorithm="HS256",
+        headers=None,
+        lifetime=None,
     ):
         """Return payload encoded to a jwt with secret and algorithm."""
         payload = copy.deepcopy(payload)
 
-        payload["exp"] = datetime.datetime.utcnow() + lifetime
+        if lifetime:
+            payload["exp"] = datetime.datetime.utcnow() + lifetime
+
+        return jwt.encode(payload, secret, algorithm=algorithm, headers=headers)
 
-        return jwt.encode(payload, secret, algorithm=algorithm)
+    @pytest.fixture(autouse=True)
+    def jwk_endpoint(self):
+        keys = {"keys": [{"kid": "KID", "kty": "RSA", "n": "1000", "e": "500"}]}
+
+        httpretty.register_uri("GET", "http://jwk.com", body=json.dumps(keys))
 
     @pytest.fixture()
     def jwt(self, patch):
         return patch("lms.services.jwt.jwt")
 
     @pytest.fixture
-    def svc(self):
-        return JWTService()
+    def svc(self, lti_registration_service):
+        svc = JWTService(lti_registration_service)
+        # Clear the lru_cache to make tests independent
+        svc._get_jwk_client.cache_clear()  # pylint: disable=protected-access
+        return svc
```

Comment:
If the JWT client was inlined as suggested above:

```suggestion
        return JWTService(lti_registration_service)
```